src/main.rs

```diff
@@ -464,50 +506,74 @@ impl Decoder {
             }
             Format::Tuple(fields) => {
                 let mut dfields = Vec::with_capacity(fields.len());
-                let mut fields = fields.iter().peekable();
-
-                while let Some(f) = fields.next() {
-                    let opt_next = match fields.peek() {
-                        Some(opt_next) => Some(*opt_next),
-                        None => opt_next,
+                let mut fields = fields.as_slice();
+
+                while let Some((f, remain)) = fields.split_first() {
+                    fields = remain;
+                    let df = match (remain.is_empty(), opt_next) {
+                        (true, None) => Decoder::compile(f, None)?,
+                        (true, Some(next)) => Decoder::compile(f, Some(next))?,
+                        (false, None) => {
+                            Decoder::compile(f, Some(&Format::Tuple(remain.to_vec())))?
+                        }
+                        (false, Some(next)) => Decoder::compile(
+                            f,
+                            Some(&Format::Cat(
+                                Box::new(Format::Tuple(remain.to_vec())),
+                                Box::new(next.clone()),
+                            )),
+                        )?,
                     };
-                    let df = Decoder::compile(f, opt_next)?;
                     dfields.push(df);
                 }
```

Comment:
Might be a little cleaner to use [`std::slice::Iter::as_slice`](https://doc.rust-lang.org/std/slice/struct.Iter.html#method.as_slice):

```rust
let mut dfields = Vec::with_capacity(fields.len());
let mut fields = fields.iter();

while let Some(f) = fields.next() {
    let df = match (remain.is_empty(), opt_next) {
        (true, None) => Decoder::compile(f, None)?,
        (true, Some(next)) => Decoder::compile(f, Some(next))?,
        (false, None) => {
            Decoder::compile(f, Some(&Format::Tuple(fields.as_slice().to_vec())))?
        }
        (false, Some(next)) => Decoder::compile(
            f,
            Some(&Format::Cat(
                Box::new(Format::Tuple(fields.as_slice().to_vec())),
                Box::new(next.clone()),
            )),
        )?,
      };
      dfields.push(df);
}
```

Reply:
Would the `remain.is_empty()` need to be a peek or is there some other way to check if it's the last one? Or just `fields.as_slice().is_empty()` :thinking:

Reply:
Yeah I think that’s `.as_slice().is_empty()`. Maybe not as clean I dunno - and to be honest I think `Iter::as_slice` could have a clearer name.

Reply:
Oh, you could just use `fields.is_empty()` (from `ExactSizeIterator`)

Reply:
Looks like that's still unstable, but len() == 0 works.